### ISSUE TITLE
[feature]: Locator: Support dependency with signalReady and dependsOn

### ISSUE THREAD

**Issue by jakusb** [Stacked-Org/stacked#1071]
### Is your feature request related to a problem? Please describe.

It does not seem possible to specify Locator to wait for other Singleton to be loaded fully first.
Example in app.dart:
  dependencies: [
    LazySingleton(classType: BottomSheetService),
    LazySingleton(classType: DialogService),
    LazySingleton(classType: RouterService),
    LazySingleton(classType: FirebaseHelper),
    LazySingleton(classType: AuthCubit),
    LazySingleton(classType: T2OUserProfileCubit),
    LazySingleton(classType: T2OThemeCubit),
    LazySingleton(classType: InternetCubit),

this results in app.locator.dart:
// Register dependencies
  locator.registerLazySingleton(() => BottomSheetService());
  locator.registerLazySingleton(() => DialogService());
  locator.registerLazySingleton(() => RouterService());
  locator.registerLazySingleton(() => FirebaseHelper());
  locator.registerLazySingleton(() => AuthCubit());
  locator.registerLazySingleton(() => T2OUserProfileCubit());
  locator.registerLazySingleton(() => T2OThemeCubit());
  locator.registerLazySingleton(() => InternetCubit());

But this is what I need. I implemented signals in my Cubits
// Register dependencies
  locator.registerLazySingleton(() => BottomSheetService());
  locator.registerLazySingleton(() => DialogService());
  locator.registerLazySingleton(() => RouterService());
  locator.registerSingleton(FirebaseHelper(),signalsReady: true);
  await locator.allReady();
  locator.registerSingletonWithDependencies(() => AuthCubit(),dependsOn: [FirebaseHelper], signalsReady: true);
  await locator.allReady();
  locator.registerSingletonWithDependencies(() => T2OUserProfileCubit(),dependsOn: [AuthCubit]);
  locator.registerSingletonWithDependencies(() => T2OThemeCubit(),dependsOn: [T2OUserProfileCubit]);
  locator.registerSingleton(InternetCubit());
  await locator.allReady();

### Describe the solution you would like

Something like:     LazySingletonWithDependencies(classType: T2OThemeCubit, dependsOn: [T2OUserProfileCubit], signalsReady: true , waitForSignalReady: true )

**Comment by ferrarafer** [comment]
Hi @jakusb , are you using Stacked and Cubit at the same time? Why are you doing that ? Stacked offers a built-in state management solution.

**Comment by jakusb** [comment]
> Hi @jakusb , are you using Stacked and Cubit at the same time? Why are you doing that ? Stacked offers a built-in state management solution.

Hi @ferrarafer , reason is that I am migrating to Stacked framework and have existing cubits.
Also I like the cubit approach. I like the BlocBuilder to update only small parts of UI and how cubits make it really modular. A properly defined cubit can be easily be used in any other project.
The reason I mainly need dependencies loading in sequence instead of parallel is to properly wait for FirebaseAuth to fully properly load, before other cubits start querying firestore. 
Also I need to decide if an anonymous login is required. I can only do that if I am certain FirebaseAuth is not logging in automatically from credentials it stored locally.
Anyway, I now have to work around and restore my app.locator file everytime stacked generate is run.

**Comment by FilledStacks** [comment]
@jakusb You can wait for async singletons with the `InitializableSingleton` type. 

```dart
InitializableSingleton(
  classType: AppInfoService,
),
```

Class definition
```dart
class AppInfoService implements InitializableDependency {
  final _firestoreApi = locator<FirebaseFirestoreApi>();

  late AppInfo appInfo;

  @override
  Future<void> init() async {
    appInfo = await _firestoreApi.getAppInformation();
  }
}

```

Generate code will look like this:
```dart
 final sweetCore = await SweetCore.setup();
  locator.registerSingleton(sweetCore);

  final appInfoService = AppInfoService();
  await appInfoService.init(); <==== here we wait until complete before moving on
  locator.registerSingleton(appInfoService);

  locator
      .registerLazySingleton<AnalyticsServiceBaseClass>(() => PosthogService());
  locator.registerLazySingleton(() => VersionService());
```

Hope that helps.

**Comment by FilledStacks** [comment]
More docs [here](https://stacked.filledstacks.com/docs/in-depth/service-locator#initializablesingleton)

**Comment by jakusb** [comment]
> More docs [here](https://stacked.filledstacks.com/docs/in-depth/service-locator#initializablesingleton)

Great! These docs were helpful, but would benefit greatly from above code example. Please consider adding them to doc. 
Also section header texts are a bit confusing. Please consider making them a bit less ambiguous. :)
Lastly: Keep doing what you are doing. Stacked is a great foundation framework for building new apps on.
I solved exactly the nasty repetitive startup trouble we experienced with every now app.

**Comment by ferrarafer** [comment]
>  I like the BlocBuilder to update only small parts of UI and how cubits make it really modular. A properly defined cubit can be easily be used in any other project.

@jakusb can you please expand on that, could be useful when we plan new features or improvements for Stacked framework.

**Comment by jakusb** [comment]
> > I like the BlocBuilder to update only small parts of UI and how cubits make it really modular. A properly defined cubit can be easily be used in any other project.
> 
> @jakusb can you please expand on that, could be useful when we plan new features or improvements for Stacked framework.

Of course.
We have tried several State Management approaches, first Redux, but that was quickly abandoned when we tried out Bloc framework. Main reason is that Redux was confusing and messy and every addition was needed in multiple places.
Bloc makes it much easier (we use Cubits, not original Blocs), to keep specific functions defined per Cubit.
So we have an Internet Cubit that simply sends events when Network state changes (none, wifi, cellular, etc).
And We have a Cubit that manages the Firebase Authentication state. Also 1 cubit per collection in Firebase. 
Anyway, a properly designed cubit can be easily reused in many projects.
Simply register the Cubit in top level in the app tree and use BlocBuilders and BlocListeners to process events.
BlocBuilder to update a portion of the UI, instead of whole pages.
BlocListener to listen for events and only update UI under specific conditions.
We love the GetIt Locator to simply get services, but it does not seem to enable updating only certain parts of ui.
Imagine a large table (30x100 cells) Every Cell an UI element. You do not want to redraw all 3000 cells, when only 1 needs updating.
